course2/public/guess.js

```diff
@@ -0,0 +1,206 @@
+			} else {
+				wordHtml += '<span class="separator">-</span>';
+			}
+		});
+		
+		wordTag.innerHTML = wordHtml;
+		
+		//Keep it globally for other functions
+		chosenWord.push(wordToGuess);
+	};
+
+	let gameOver = function(status){
+
+		$("#inputLetter").attr('readOnly', 'readOnly');
+		switch (status){
+			case true:
+				$('#mainMsg').text('Congrats, you win!');
+				break;
+			case false:
+				$('#mainMsg').text('You loose! Try again');
+				break;
+		}
+
+		$('#gameReset').on('click', function(){
+			location.reload();
+		});
+	}
+
+	let guessMe = function(letter){
+		let regex = new RegExp(letter, 'gi'),
+			matches = chosenWord[0].toString().match(regex);
+
+		return matches ? matches.length : 0;
+	}
+
+	let updateGuessingWord = function(letters){
+
+		let wordToBeGuessed = [...chosenWord[0].toString()],
+			finalWord = [];
+
+		wordToBeGuessed.map(function(value){
+			value = value.toLowerCase();
+			if(letters.indexOf(value) !== -1){
+				finalWord.push(value)
+			}else if(value == ','){
+				finalWord.push('-')
+			}else{
+				finalWord.push('.')
+			}
+		});
+
+		if(finalWord.indexOf('.') == -1){
+			gameOver(true);
+		}
+
+		return finalWord;
+	}
+
+	let markupUpdate = function(content){
+		let html = '';
+
+		content.map(function(value){
+			if(value == '-'){
+				html += `<span class="separator">-</span>`;
+			}else if(value !== '.'){
+				html += `<span>${value}</span>`
+			}else{
+				html += `<span>.</span>`
+			}
+		});
+
+		$('#wordToGuess').html(html);
+	}
+
+	let isValidLetter = function (letter) {
+		let errCount = 0,
+			errMsg = [],
+			validLetters = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z'];
+
+		if (letter == '') {
+			errCount += 1;
+			errMsg.push('You have to provide a letter first');
+		}else if(validLetters.indexOf(letter) == -1){
+			errCount += 1;
+			errMsg.push('Only letters are allowed');
+		}
+		return [errMsg, errCount];
+	};
+
+	let guessLetter = function(){
+
+		let guessedLetters = [],
+			failAttemps = [],
+			foundedLetters,
+			bonus = 0;
+		$('#staticTries').val(MAX_FAILS);
+
+		$("#inputLetter").on('keyup', function(){
+
+			let currentLetter = "";
+			$('#errorMsg').text('');
+
+			/*When typing fast - sometimes an empty value kiks in*/
+			if($(this).val() !== ""){
+				currentLetter = $(this).val();
+			}
+
+			if(isValidLetter(currentLetter)[0].length > 0) {
+				$('#errorMsg').text(isValidLetter(currentLetter)[0].toString());
+				$(this).val('')
+				return;
+			}else {
+
+				let match = guessMe(currentLetter);
+
+				if (match > 0) {
+					if (guessedLetters.indexOf(currentLetter) == -1) {
+						guessedLetters.push(currentLetter);
+						foundedLetters = updateGuessingWord(guessedLetters);
+						bonus += (match * BONUS_PER_LETTER);
+						$('#errorMsg').text('');
+					} else {
+						$('#errorMsg').text('You already entered that letter')
+					}
+				} else {
+					failAttemps.push(currentLetter);
```

Comment:
I like this idea of keeping track of entered letters, but it can be applied so that a duplicate letter fail doesn't count as a fail. Eg: If I enter `w` twice and it's not in the word, it would be nice to not decrease the fail count and just say the above message "You already entered that letter".

Reply:
Done!